gateway/gateway.go

```diff
@@ -2,36 +2,48 @@ package gateway
 
 import (
 	"context"
+	"io"
 	"net/http"
 	"sort"
 
-	coreiface "github.com/ipfs/interface-go-ipfs-core"
-	path "github.com/ipfs/interface-go-ipfs-core/path"
+	cid "github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-libipfs/files"
+	iface "github.com/ipfs/interface-go-ipfs-core"
+	options "github.com/ipfs/interface-go-ipfs-core/options"
+	"github.com/ipfs/interface-go-ipfs-core/path"
 )
 
 // Config is the configuration that will be applied when creating a new gateway
 // handler.
 type Config struct {
-	Headers  map[string][]string
-	Writable bool
+	Headers map[string][]string
 }
 
-// NodeAPI defines the minimal set of API services required by a gateway handler
+// NodeAPI defines the minimal set of API services required for a read-only
+// gateway handler. These are, for the most part, subsets of the core interface.
+
 type NodeAPI interface {
-	// Unixfs returns an implementation of Unixfs API
-	Unixfs() coreiface.UnixfsAPI
+	// GetUnixFs returns a read-only handle to a file tree referenced by a path.
+	GetUnixFs(context.Context, path.Path) (files.Node, error)
+
+	// LsUnixFs returns the list of links in a directory.
+	LsUnixFs(context.Context, path.Path, ...options.UnixfsLsOption) (<-chan iface.DirEntry, error)
+
+	// GetBlock attempts to resolve the path and return a reader for data in the block.
+	GetBlock(context.Context, path.Path) (io.Reader, error)
 
-	// Block returns an implementation of Block API
-	Block() coreiface.BlockAPI
+	// StatBlockOffline returns information on a block without making any network
+	// connection. In case the block does not exist locally, an error should be returned.
+	StatBlockOffline(context.Context, path.Path) (iface.BlockStat, error)
 
-	// Dag returns an implementation of Dag API
-	Dag() coreiface.APIDagService
+	// GetNode returns the IPLD node for the path.
+	GetNode(context.Context, cid.Cid) (ipld.Node, error)
 
-	// Routing returns an implementation of Routing API.
-	// Used for returning signed IPNS records, see IPIP-0328
-	Routing() coreiface.RoutingAPI
+	// GetRouting retrieves the best value for a given key from the routing system.
+	GetRouting(context.Context, string) ([]byte, error)
```

Comment:
We only use this for IPNS, make it explicit, and require a valid Cid instead of a string:

```suggestion
	// GetIPNSRecord retrieves the best IPNS record for a given CID (libp2p-key) from the routing system.
	GetIPNSRecord(context.Context, cid.Cid) ([]byte, error)
```